#Contents List
##Import Libraries
##Import Data - orders_products_merged.pkl
##Create a user-defined function
##Applying the function
##If statement with loc() function
##If statement with loc() function on df_ords_prods_merge dataframe
##Frequency of the orders_day_of_week column
##Creating the for-loop
##Combining with the df_ords_prods_merge dataframe
##2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.
##Combine results with the df_ords_prods_merge dataframe
##3.Check Values for accuracy
##4 When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”
##Create a for-loop code assigning time period labels
##Combine results with the df_ords_prods_merge dataframe
##5. Print the frequency for this new column.
##7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder. - orders_products_merged_updated.pkl

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Family Computer\Documents\Data Analytics\Instacart Basket Analysis'

In [3]:
#Import ords_prods_merge
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [5]:
df_ords_prods_merge.shape

(32404859, 14)

In [6]:
df = df_ords_prods_merge[:1000000]

In [7]:
df.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [8]:
df.shape

(1000000, 14)

In [9]:
#Create a user-defined function
def price_label(row):
  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High-range'
  else: return 'Not enough data'

In [10]:
#Applying the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Family Computer\AppData\Local\Temp\ipykernel_19552\3087574758.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [11]:
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [12]:
df['prices'].max()

14.8

In [13]:
#If statement with loc() function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Family Computer\AppData\Local\Temp\ipykernel_19552\2439778038.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [17]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   2539329        1             1                   2                    8   
1   2398795        1             2                   3                    7   
2    473747        1             3                   3                   12   
3   2254736        1             4                   4                    7   
4    431534        1             5                   4                   15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [18]:
df_ords_prods_merge.shape

(32404859, 14)

In [27]:
#If statement with loc() function on df_ords_prods_merge dataframe
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [30]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [31]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [32]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [33]:
#Frequency of the orders_day_of_week column
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [34]:
#Creating the for-loop
result = []
for value in df_ords_prods_merge['orders_day_of_week']:
    if value ==0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regular busy')

In [35]:
result

['Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Regular busy',
 'Busiest day',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular bu

In [36]:
#Combining with the df_ords_prods_merge dataframe
df_ords_prods_merge['busiest day'] = result

In [37]:
df_ords_prods_merge['busiest day'].value_counts (dropna = False)

busiest day
Regular busy    22416875
Busiest day      6204182
Least busy       3783802
Name: count, dtype: int64

In [49]:
#2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.
result = []
for value in df_ords_prods_merge['orders_day_of_week']:
    if value == 0 or value == 1:
        result.append('Busiest days')
    elif value == 3 or value == 4:
        result.append('Slowest days')
    else:
        result.append('Regular busy')

In [50]:
result

['Regular busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Slowest days',
 'Regular busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Slowest days',
 'Slowest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Slowest days

In [51]:
#Combine results with the df_ords_prods_merge dataframe
df_ords_prods_merge['busiest day'] = result

In [53]:
#3.Check Values for accuracy
df_ords_prods_merge['busiest day'].value_counts (dropna = False)

busiest day
Regular busy    12916111
Busiest days    11864412
Slowest days     7624336
Name: count, dtype: int64

##3. Note any observatrions - The total for both for-loops is 32,404,859, so all rows have been accounted for. The numbers have been adjusted accourdingly to account for the new Busiest and slowest day headings. 

In [54]:
#4 When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”
df_ords_prods_merge['hour_of_day_ordered'].value_counts (dropna = False)

hour_of_day_ordered
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [60]:
#Create a for-loop code assigning time period labels
result = []
for value in df_ords_prods_merge ['hour_of_day_ordered']:
    if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9:
        result.append('Most orders')
    elif value == 17 or value == 8 or value == 18 or value == 19 or value == 20 or value == 7 or value == 21 or value == 22:
        result.append('Average orders')
    else:
        result.append('Fewest orders')

In [61]:
result

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [62]:
##Combine results with the df_ords_prods_merge dataframe
df_ords_prods_merge['busiest_period_of_day'] = result

In [63]:
#5. Print the frequency for this new column.
df_ords_prods_merge['busiest_period_of_day'].value_counts (dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

##5. All headings add up to 32,404,859.

In [65]:
#7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))